### PR TITLE
Adding volume resource creation to vSphere provider

### PR DESCRIPTION
**Pull request by dkalleg**
Allows the user to create a vmdk in vSphere to the given path. Update
not implemented, further research required.  In the future this could
support updates for rename and move.

### CODE REVIEW

**Comment by chrislovecnm** [comment]
@dkalleg is this fixing an open issue? A bit brain dead this morning ;) I am thinking that this is a new feature?

**Comment by dkalleg** [comment]
New feature

File: builtin/providers/vsphere/resource_vsphere_virtual_disk.go [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+	if err != nil {
+		return err
+	}
+
+	fileInfo, err := ds.Stat(context.TODO(), vDisk.vmdkPath)
+	if err != nil {
+		log.Printf("[DEBUG] resourceVSphereVirtualDiskRead - stat failed on: %v", vDisk.vmdkPath)
+		d.SetId("")
+		return err
+	} else {
```
Comment (chrislovecnm) [inline builtin/providers/vsphere/resource_vsphere_virtual_disk.go]: Don't think we need this else

File: builtin/providers/vsphere/resource_vsphere_virtual_disk.go [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+
+type virtualDisk struct {
+	size       int
+	vmdkPath   string
+	initType   string
+	datacenter string
+	datastore  string
+}
+
+func resourceVSphereVirtualDisk() *schema.Resource {
```
Comment (chrislovecnm) [inline builtin/providers/vsphere/resource_vsphere_virtual_disk.go]: Method comments ... Pretty please. What does this do ..

File: builtin/providers/vsphere/resource_vsphere_virtual_disk.go [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+
+	diskPath := fmt.Sprintf("[%v] %v", vDisk.datastore, vDisk.vmdkPath)
+
+	err := createHardDisk(client, vDisk.size, diskPath, vDisk.initType, vDisk.datacenter)
+	if err != nil {
+		return err
+	}
+
+	d.SetId(diskPath)
+
```
Comment (chrislovecnm) [inline builtin/providers/vsphere/resource_vsphere_virtual_disk.go]: Should we have some logging?  In general a touch more logging would be nice ;)

File: builtin/providers/vsphere/resource_vsphere_virtual_disk.go [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+	dc, err := getDatacenter(client, d.Get("datacenter").(string))
+	if err != nil {
+		return err
+	}
+	diskPath := fmt.Sprintf("[%v] %v", vDisk.datastore, vDisk.vmdkPath)
+
+	virtualDiskManager := object.NewVirtualDiskManager(client.Client)
+
+	virtualDiskManager.DeleteVirtualDisk(context.TODO(), diskPath, dc)
```
Comment (chrislovecnm) [inline builtin/providers/vsphere/resource_vsphere_virtual_disk.go]: Does the API return an error on this call if it fails? For instance we try to delete a disk that is not there.

File: builtin/providers/vsphere/resource_vsphere_virtual_disk.go [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+	d.SetId("")
+	return nil
+}
+
+// createHardDisk creates a new Hard Disk.
+func createHardDisk(client *govmomi.Client, size int, diskPath string, diskType string, dc string) error {
+	virtualDiskManager := object.NewVirtualDiskManager(client.Client)
+	spec := &types.FileBackedVirtualDiskSpec{
+		VirtualDiskSpec: types.VirtualDiskSpec{
+			AdapterType: "ide",
```
Comment (chrislovecnm) [inline builtin/providers/vsphere/resource_vsphere_virtual_disk.go]: Are there other adapter types?

File: builtin/providers/vsphere/resource_vsphere_virtual_disk_test.go [hunk 6]
```diff
@@ -0,0 +1,125 @@
+package vsphere
```
Comment (chrislovecnm) [inline builtin/providers/vsphere/resource_vsphere_virtual_disk_test.go]: Does the TF framework support negative test cases?

**Comment by chrislovecnm** [comment]
Appreciate the work. Looking forward to getting this in. Got a bunch of comments for yah ;)

File: builtin/providers/vsphere/resource_vsphere_virtual_disk_test.go (hunk 6 above)
Comment (dkalleg) [inline builtin/providers/vsphere/resource_vsphere_virtual_disk_test.go]: Looking at other tests and at helper/resource/testing.go library, it looks like negative tests aren't supported.

File: builtin/providers/vsphere/resource_vsphere_virtual_disk.go (hunk 5 above)
Comment (dkalleg) [inline builtin/providers/vsphere/resource_vsphere_virtual_disk.go]: Yep, I'll add an option for them.

**Comment by dkalleg** [comment]
@chrislovecnm @kristinn @phinze @tkak Please give this a review at your earliest convenience.

File: builtin/providers/vsphere/resource_vsphere_virtual_disk_test.go (hunk 6 above)
Comment (chrislovecnm) [inline builtin/providers/vsphere/resource_vsphere_virtual_disk_test.go]: @stack72 @jen20 @phinze say it ain't so ... no negative testing??  I will file an issues

File: builtin/providers/vsphere/resource_vsphere_virtual_disk_test.go [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+
+		finder = finder.SetDatacenter(dc)
+
+		ds, err := finder.Datastore(context.TODO(), rs.Primary.Attributes["datastore"])
+		if err != nil {
+			return fmt.Errorf("error %s", err)
+		}
+
+		fileInfo, err := ds.Stat(context.TODO(), rs.Primary.Attributes["vmdk_path"])
+		_ = fileInfo
```
Comment (kristinn) [inline builtin/providers/vsphere/resource_vsphere_virtual_disk_test.go]: What's the reason for this?

**Comment by kristinn** [comment]
This is a very good PR 👍 Just missing some function signatures, that's all.

File: builtin/providers/vsphere/resource_vsphere_virtual_disk_test.go (hunk 7 above)
Comment (dkalleg) [inline builtin/providers/vsphere/resource_vsphere_virtual_disk_test.go]: This is a go-ism I wasn't sure the best way to work around :)  As I started explaining myself here I realized my mistake!  I'll push up the change!

Basically, I wanted to say `_, err := ds.Stat(...` but was getting the error of 'no new vars on left of :=', so I gave it `fileInfo`.  But with that, `fileInfo` has to be used once or the go compiler will get angry, and this is essentially a noop to work around that.  All this goes away if i take out the colon :)  Doh!

Comment (chrislovecnm) [inline builtin/providers/vsphere/resource_vsphere_virtual_disk_test.go]: @stack72 @phinze @jen20 any opinion on this?  Your GoFu is strong...

Comment (stack72) [inline builtin/providers/vsphere/resource_vsphere_virtual_disk_test.go]: you can do it as follows:

```
_, err = ds.Stat(context.TODO(), rs.Primary.Attributes["vmdk_path"])
```

have already declared `err` above AFAICT.

Comment (dkalleg) [inline builtin/providers/vsphere/resource_vsphere_virtual_disk_test.go]: Sorry, didn't mean to pose that as a question.  Thats the answer that came to me after reading kristinn's question :)

Updating the PR shortly

File: builtin/providers/vsphere/resource_vsphere_virtual_disk_test.go (hunk 6 above)
Comment (jen20) [inline builtin/providers/vsphere/resource_vsphere_virtual_disk_test.go]: Not sure I follow what you mean by negative testing?

File: builtin/providers/vsphere/resource_vsphere_virtual_disk.go [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+
+	finder := find.NewFinder(client.Client, true)
+	finder = finder.SetDatacenter(dc)
+
+	ds, err := finder.Datastore(context.TODO(), d.Get("datastore").(string))
+	if err != nil {
+		return err
+	}
+
+	fileInfo, err := ds.Stat(context.TODO(), vDisk.vmdkPath)
```
Comment (dagnello) [inline builtin/providers/vsphere/resource_vsphere_virtual_disk.go]: if you don't need fileInfo, you can replace it with:
_, err = ds.Stat(context.TODO(), vDisk.vmdkPath)

File: builtin/providers/vsphere/resource_vsphere_virtual_disk_test.go [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+		}
+		finder = finder.SetDatacenter(dc)
+
+		ds, err := finder.Datastore(context.TODO(), rs.Primary.Attributes["datastore"])
+		if err != nil {
+			return fmt.Errorf("error %s", err)
+		}
+
+		fileInfo, err := ds.Stat(context.TODO(), rs.Primary.Attributes["vmdk_path"])
+		_ = fileInfo
```
Comment (dagnello) [inline builtin/providers/vsphere/resource_vsphere_virtual_disk_test.go]: if you don't need fileInfo, you can replace it with:
_, err = ds.Stat(context.TODO(), vDisk.vmdkPath)

File: builtin/providers/vsphere/resource_vsphere_virtual_disk_test.go (hunk 7 above)
Comment (dagnello) [inline builtin/providers/vsphere/resource_vsphere_virtual_disk_test.go]: same as above

Comment (chrislovecnm) [inline builtin/providers/vsphere/resource_vsphere_virtual_disk_test.go]: @dagnello thanks for the help!

File: website/source/docs/providers/vsphere/r/virtual_disk.html.markdown [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+```
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `size` - (Required) Size of the disk (in GB).
+* `vmdk_path` - (Required) The path, including filename, of the virtual disk to be created.  This should end with '.vmdk'.
+* `type` - (Optional) 'eagerZeroedThick' (the default), or 'thin' are supported options.
+* `datacenter` - (Required) The name of a Datacenter in which to create the disk.
```
Comment (dagnello) [inline website/source/docs/providers/vsphere/r/virtual_disk.html.markdown]: I think this should be optional, similar to folder or virtual machine resources

**Comment by dkalleg** [comment]
@kristinn Could you point out which function signature you want to see updated?  It isn't obvious for me.  Thx.

File: builtin/providers/vsphere/resource_vsphere_virtual_disk.go [hunk 11]
```diff
@@ -0,0 +1,250 @@
...
+		Create: resourceVSphereVirtualDiskCreate,
+		Read:   resourceVSphereVirtualDiskRead,
+		Delete: resourceVSphereVirtualDiskDelete,
+
+		Schema: map[string]*schema.Schema{
+			// Size in GB
+			"size": &schema.Schema{
+				Type:     schema.TypeInt,
+				Required: true,
+				ForceNew: true, //TODO Can this be optional (resize)?
```
Comment (dagnello) [inline builtin/providers/vsphere/resource_vsphere_virtual_disk.go]: it looks like these can be false if an upgrade supports the change

Comment (dkalleg) [inline builtin/providers/vsphere/resource_vsphere_virtual_disk.go]: That would be changed in a PR that also adds vm Update functionality.  For now it needs to stay this way, else TF would go looking for a `resourceVSphereVirtualMachineUpdate` that isn't there.

**Comment by dkalleg** [comment]
@chrislovecnm @phinze @tkak Any more feedback?

**Comment by chrislovecnm** [comment]
LGTM!

**Comment by stack72** [comment]
The main point I have here is that we are not setting the state of any values on the read. Its a normal practice to use the `d.Set("param", value)` notation

**Comment by dkalleg** [comment]
@stack72  Ah yes, let me add that and I'll push an update to this PR hopefully sometime today or tomorrow.  Since we aren't getting any specific vmdk info* on the read, I won't be able to fetch every value.  But should be able to do a d.Set for size, path, datacenter?, datastore.  But not type and adapter type.

*I didn't see any way in govmomi apis to drill down and get vmdk-level data on a file, like init_type and adapter_type.  If anyone knows how to do this I'd like to be able to add that functionality!  Maybe it required changes to govmomi.

File: builtin/providers/vsphere/resource_vsphere_virtual_disk.go [hunk 12]
```diff
@@ -0,0 +1,250 @@
...
+func resourceVSphereVirtualDiskCreate(d *schema.ResourceData, meta interface{}) error {
+	log.Printf("[INFO] Creating Virtual Disk")
+	client := meta.(*govmomi.Client)
+
+	vDisk := virtualDisk{
+		size: d.Get("size").(int),
+	}
+
+	if v, ok := d.GetOk("vmdk_path"); ok {
+		vDisk.vmdkPath = v.(string)
```
Comment (chrislovecnm) [inline builtin/providers/vsphere/resource_vsphere_virtual_disk.go]: @stack72 you would expect the `set` usage here?  More info please.

Comment (stack72) [inline builtin/providers/vsphere/resource_vsphere_virtual_disk.go]: @chrislovecnm the setting of the state usually happens I'm the read func so we are sure that we expected to happen has happened.

**Comment by chrislovecnm** [comment]
@stack72 you mind a concrete example?  What line in the read func would you like to be modified.  I think I understand, but I don't want to assume 😄

**Comment by dkalleg** [comment]
@stack72 I believe I've addressed your comment.  As of right now, I don't know how to query vSphere vdisks for more info other than simply stat'ing them (even the stat call doesn't seem to be returning the file size properly), so I couldn't add writes for things like size, type, ect.  But I've added a todo there to mark further investigation on the govmomi side.

Keep those reviews coming :)

**Comment by chrislovecnm** [comment]
@dkalleg should we open a ticket with govmomi peeps asking the question about vdisk attributes?

**Comment by dkalleg** [comment]
That's probably a good idea.  Even if there is some wacky way to accomplish it today, there isn't support in the object/datastore api, which I would think should have it.

**Comment by chrislovecnm** [comment]
@dkalleg make it so .. and please cc: @markpeek - he works in the same group as the govmomi guys

**Comment by dougm** [comment]
Datastore.Stat returns an interface type (`types.BaseFileInfo`), in the case of `.vmdk` files will be `*types.VmDiskFileInfo` with the fields it sounds like you're looking for:

```go
res, err := Datastore.Stat(ctx, vmdkPath)
if err != nil {...}
info, ok := res.(*types.VmDiskFileInfo)
if !ok { panic("unexpected type") }
// make use of info.{DiskType,CapacityKb,etc}
```

**Comment by chrislovecnm** [comment]
@dougm much thanks

**Comment by dkalleg** [comment]
@dougm Thanks for the tip.  When I store the result of the assertion / type-cast (I recently learned the proper term for it is a assertion, not a type-cast.. another weird go-ism) and print out its values, they're mostly default / zero'd values.  

```
	fileInfo, err := ds.Stat(context.TODO(), vDisk.vmdkPath)
	if err != nil {
		log.Printf("[DEBUG] resourceVSphereVirtualDiskRead - stat failed on: %v", vDisk.vmdkPath)
		d.SetId("")
		return err
	}
	log.Printf("[DEBUG] resourceVSphereVirtualDiskRead - fileinfo: %#v", fileInfo.GetFileInfo())

	diskInfo := fileInfo.(*types.VmDiskFileInfo)
	log.Printf("[DEBUG] resourceVSphereVirtualDiskRead - disk info: %#v", diskInfo)
```

```
2016/05/09 16:26:45 [DEBUG] terraform-provider-vsphere: 2016/05/09 16:26:45 [DEBUG] resourceVSphereVirtualDiskRead - fileinfo: &types.FileInfo{DynamicData:types.DynamicData{}, Path:"newVmdk.vmdk", FileSize:0, Modification:(*time.Time)(nil), Owner:"root"}
2016/05/09 16:26:45 [DEBUG] terraform-provider-vsphere: 2016/05/09 16:26:45 [DEBUG] resourceVSphereVirtualDiskRead - disk info: &types.VmDiskFileInfo{FileInfo:types.FileInfo{DynamicData:types.DynamicData{}, Path:"newVmdk.vmdk", FileSize:0, Modification:(*time.Time)(nil), Owner:"root"}, DiskType:"", CapacityKb:0, HardwareVersion:0, ControllerType:"", DiskExtents:[]string(nil), Thin:(*bool)(nil)}
```

Focusing on just one attribute, I can confirm the disk is 2GB, but the returned info suggests its 0kb.  Any suggestions or ideas on why Stat is giving me this zero'd data?

**Comment by dougm** [comment]
@dkalleg this PR will fix that: https://github.com/vmware/govmomi/pull/509

Should get that merged later or tomorrow.

**Comment by dkalleg** [comment]
@dougm Thanks so much for that lightning fast turn around!  Does this mean I will continue to expect the remaining fields to not update?  ex. ControllerType, Thin, CapacityKb, HardwareVersion, ect

```
type VmDiskFileInfo struct {
	FileInfo

	DiskType        string   `xml:"diskType,omitempty"`
	CapacityKb      int64    `xml:"capacityKb,omitempty"`
	HardwareVersion int32    `xml:"hardwareVersion,omitempty"`
	ControllerType  string   `xml:"controllerType,omitempty"`
	DiskExtents     []string `xml:"diskExtents,omitempty"`
	Thin            *bool    `xml:"thin"`
}
```

@chrislovecnm @stack72 What this means now, is:
If we want this PR to include data on size/initType that the PR trying to update our govmomi version needs to update far enough to include Doug's fix.
or
Merge what we have with the todo of updating it once govmomi v0.6.0+ lands.

**Comment by dkalleg** [comment]
@dougm Unrelated question.. Can you explain about types.DynamicData?  I see the VirtualDevice struct extends DynamicData, which is an empty struct.  What is its purpose?  Is it a way for govmomi users to add in dynamic data to be associated with a VirtualDevice in vSphere?  Or is it a way for vSphere to send non-api defined dynamic data back to the govmomi user?  Or something else?  Feel free to email me at dkalleg at vt dot edu

**Comment by chrislovecnm** [comment]
@dkalleg you mind putting in a issue with govmomi? ... Just for tracking.

Thanks

Chris

**Comment by dkalleg** [comment]
@chrislovecnm @stack72 Any further comments on this PR?

**Comment by chrislovecnm** [comment]
@stack72 @jen20 @phinze can we get a final review and then have it merge?

**Comment by chrislovecnm** [comment]
@dkalleg which version of govmomi do you need for this??

**Comment by dkalleg** [comment]
This was built on top of the old 0.3.0.  I'll need to rebase this.  Working on vm resource disk update story today so I might not get to this til next week.

**Comment by chrislovecnm** [comment]
@dkalleg can you mark this [WIP].  Appreciate the work! We probably going to bump the version again.

**Comment by dkalleg** [comment]
@chrislovecnm @stack72 I've updated to include the new size info via our govmomi upgrade with Doug's updated Stat() info.  Please review at your earliest convenience.

**Comment by chrislovecnm** [comment]
@dkalleg is this good to go?

**Comment by dkalleg** [comment]
@chrislovecnm I believe so

**Comment by chrislovecnm** [comment]
@catsby @jen20 @phinze @stack72 can we get a merge!

**Comment by chrislovecnm** [comment]
@jen20 @stack72 @phinze @catsby how do we get this merged? Help is appreciated!!

**Comment by stack72** [comment]
LGTM now :) Thanks for the extra additions as requested

P.

**Comment by dkalleg** [comment]
Thanks @stack72 !

**Pull request merged**